toffy/streak_detection.py

```diff
@@ -0,0 +1,410 @@
+    mean of pixel above and below it.
+
+    Args:
+        input_image (np.ndarray): The input image (Numpy array representation of the tiff file) to
+        be used for streak correction.
+        min_row (int): The minimum row index of the streak. The y location where the streak
+        starts.
+        max_row (int): The maximum row index of the streak. The y location where the streak ends.
+        min_col (int): The minimum column index of the streak. The x location where the streak
+        starts.
+        max_col (int): The maximum column index of the streak. The x location where the streak
+        ends.
+
+    Returns:
+        np.ndarray: Returns the corrected streak.
+    """
+    streak_corrected: np.ndarray = np.mean(
+        [
+            # Row above
+            input_image[min_row, min_col + 1: max_col + 1],
+            # Row below
+            input_image[max_row + 1, min_col + 1: max_col + 1],
+        ],
+        axis=0,
+        dtype=np.int8,
+    )
+
+    return streak_corrected
+
+
+def streak_correction(
+    fov_dir: Union[str, Path],
+    image_name: str = "Noodle",
+    mask_data: bool = False,
+) -> Optional[StreakData]:
+    """Takes a `fov` directory and a user specified image for streak detection. Once all the
+    streaks have been detected on that image, they are corrected via an averaging method. The
+    function can also return a DataClass containing various binary masks and dataframes which were
+    used for filtering and correction when prompted to.
+
+    Args:
+        fov_dir (Union[str, Path]): A directory containing the fov and all it's channels for
+        correction.
+        image_name (str): The name of the image used for identifying and removing the streaks.
+        Defaults to "Noodle".
+        mask_data (bool): If `True`, returns a DataClass consisting of binary masks and dataframes
+        for the detected streaks. Defaults to "False"
+
+    Returns:
+        Optional[StreakData]: A DataClass holding all necessary data for streak detection and
+        correction as well as masks useful for debugging and visualization.
+    """
+
+    # Initialize the streak DataClass
+    streak_data = StreakData(streak_channel=image_name)
+
+    # Open fov directory containing all the tiff files
+    fov_data = load_utils.load_imgs_from_dir(data_dir=fov_dir, dtype=np.int32)
+
+    #  Open the image for mask generation.
+    with fov_data.sel(fovs=image_name) as data:
+        # Reshape it from (x,x,1) to (x,x)
+        input_image = data.values.reshape(2048, -1)
+        streak_data.shape = input_image.shape
+        # Create and filter the binary masks
+        streak_data.streak_mask = _make_binary_mask(input_image=input_image)
+        _filter_mask(streak_data=streak_data, min_length=50)
+
+    # Get the file names.
+    channel_fn = fov_data.fovs.values.tolist()
+
+    # Initialize the corrected images.
```

Comment:
Rather than a dict, let's use an xarray to hold the corrected images. You can initialize with an empty numpy array